### DISCUSSION TITLE
Dependency Injection Framework

### DISCUSSION THREAD

**Discussion by Czeach** [AndroidDev-social/DodoForMastodon#51] [General]
### Description

Set up Hilt (or Dagger) library for dependency injection

**Comment by blvckcoded** [comment]
Should we consider Koin instead? If we truly want to go multiplatform? There are performance tradeoffs between these DI frameworks.

Build times will be faster with Koin but that shouldn't be the sole reason for it... Let's chat more about this 😃

**Comment by evant** [comment]
If you want something like dagger that supports multiplatform 😉 https://github.com/evant/kotlin-inject

**Reply by blvckcoded** [reply]
I love it 😂😂😂

**Comment by svenjacobs** [comment]
Hm, Hilt / Dagger would only be possible for Android. But this is a multiplatform project. I personally think the project should just use custom DI. Using the same solution for the whole project, even shared code. With Kotlin's [default parameter values](https://blog.devgenius.io/all-you-need-for-dependency-injection-is-kotlin-831b235bb45b) it's pretty simple, less overhead.

**Comment by thebino** [comment]
I would recommend, using Koin. It has support for multiplatform
https://insert-koin.io/docs/reference/koin-mp/kmp

**Comment by svenjacobs** [comment]
I really don't see the benefit of using Koin. You still have to write modules. Kotlin's default parameter values solve 90% of all DI needs _imho_. In my personal project I was using Koin before, then migrated to custom DI and was saving so much lines of code 🤩

**Comment by svenjacobs** [comment]
Just have a look [here](https://github.com/svenjacobs/leon/pull/73/files), another project of mine, where I removed Hilt and replaced it with custom DI. Look at how much LOC I could save 🤩 For everything where default parameters are not sufficient, I use objects that I call containers (examples [here](https://github.com/svenjacobs/leon/blob/bd464957b643727404e68efb6ac78a0bd6dcba42/core-domain/src/main/kotlin/com/svenjacobs/app/leon/core/domain/inject/AppContainer.kt) and [here](https://github.com/svenjacobs/leon/blob/bd464957b643727404e68efb6ac78a0bd6dcba42/app/src/main/kotlin/com/svenjacobs/app/leon/startup/ContainerInitializer.kt)).

**Comment by digitalbuddha** [comment]
My vote is to use https://github.com/evant/kotlin-inject

It is to the jsr330 spec and works in kmp. Plus Evan is here to do support if it fails 😆

**Comment by bitPogo** [comment]
I vote for koin, since it is the solution I had less trouble with in different KMP/KMM project.

**Comment by kpgalligan** [comment]
I generally default to Koin for KMP, but we did the old memory model implementation of Koin, so there's kind of a historical reason. `kotlin-inject` has been on the internal research list for a while, and as Mike pointed out, if the author is here, that's a pretty interesting situation.

**Reply by kpgalligan** [reply]
(But if you're reading this Arnaud, Koin is great, just saying ;)

**Comment by thebino** [comment]
⛔️ This discussion is closed. Please vote or [check for results.](https://github.com/AndroidDev-social/MastodonCompose/discussions/54)
